src/tally.F90

```diff
@@ -2398,117 +2401,99 @@ subroutine score_surface_current(p)
-        else
-          do j = ijk0(3), ijk1(3) + 1, -1
-            ijk0(3) = j
-            if (all(ijk0 >= 1) .and. all(ijk0 <= m % dimension)) then
-              matching_bins(i_filter_surf) = OUT_BOTTOM
-              matching_bins(i_filter_mesh) = &
-                   mesh_indices_to_bin(m, ijk0)
-              filter_index = sum((matching_bins(1:size(t % filters)) - 1) &
-                   * t % stride) + 1
-!$omp atomic
-              t % results(1, filter_index) % value = &
-                   t % results(1, filter_index) % value + p % wgt
-            end if
-          end do
-        end if
-        cycle
-      elseif (x_same .and. z_same) then
-        ! Only y crossings
-        if (uvw(2) > 0) then
-          do j = ijk0(2), ijk1(2) - 1
-            ijk0(2) = j
-            if (all(ijk0 >= 1) .and. all(ijk0 <= m % dimension)) then
-              matching_bins(i_filter_surf) = OUT_FRONT
-              matching_bins(i_filter_mesh) = &
-                   mesh_indices_to_bin(m, ijk0)
-              filter_index = sum((matching_bins(1:size(t % filters)) - 1) &
-                   * t % stride) + 1
+      ! Loop over the dimensions
+      do d1 = 1, 3
+
+        ! Get the other dimensions
+        d2 = mod(d1, 3) + 1
+        d3 = mod(d1 + 1, 3) + 1
+
+        ! If cell index in dimension d1 and d2 are the same and the cells is
+        ! within the mesh bounds
+        if (ijk0(d1) == ijk1(d1) .and. ijk0(d2) == ijk1(d2)) then
+
+          ! Only d3 crossings
+          if (uvw(d3) > 0) then
+
+            ! Loop over d3 cells
+            do j = ijk0(d3), ijk1(d3) - 1
+              ijk0(d3) = j
+
+              ! Outward current on d3 max surface
+              if (all(ijk0 >= 1) .and. all(ijk0 <= m % dimension)) then
+                matching_bins(i_filter_surf) = d3 * 2
+                matching_bins(i_filter_mesh) = &
+                     mesh_indices_to_bin(m, ijk0)
+                filter_index = sum((matching_bins(1:size(t % filters)) - 1) &
+                     * t % stride) + 1
 !$omp atomic
-              t % results(1, filter_index) % value = &
-                   t % results(1, filter_index) % value + p % wgt
-            end if
-          end do
-        else
-          do j = ijk0(2), ijk1(2) + 1, -1
-            ijk0(2) = j
-            if (all(ijk0 >= 1) .and. all(ijk0 <= m % dimension)) then
-              matching_bins(i_filter_surf) = OUT_BACK
-              matching_bins(i_filter_mesh) = &
-                   mesh_indices_to_bin(m, ijk0)
-              filter_index = sum((matching_bins(1:size(t % filters)) - 1) &
-                   * t % stride) + 1
+                t % results(1, filter_index) % value = &
+                     t % results(1, filter_index) % value + p % wgt
+              end if
+
+              ! Inward current on d3 min surface
+              if (ijk0(d1) >= 1 .and. ijk0(d1) <= m % dimension(d1) .and. &
+                   ijk0(d2) >= 1 .and. ijk0(d2) <= m % dimension(d2) .and. &
```

Comment:
It is necessary to check d1 and d2 due to the case where the mesh does not span the entire geometry. The `mesh % get_mesh_indices(...)` subroutine will return mesh indices that lie outside the bounds of the mesh dimensions (either <= 0 or greater than the mesh dimension in each direction). If the start or end point of the particle lie outside the mesh, the `tally % score_surface_current(...)` will seg fault when trying to tally to surfaces not physically in the mesh.